server/svix-server/src/core/idempotency.rs

```diff
@@ -0,0 +1,615 @@
+        let headers = out.headers_mut();
+        *headers = resp_headers
+            .iter()
+            .map(|(k, v)| Ok((k.parse()?, http::HeaderValue::from_bytes(v)?)))
+            .collect::<Result<_, ConversionToResponseError>>()?;
+    }
+
+    Ok(out)
+}
+
+struct IdempotencyKey(String);
+
+impl IdempotencyKey {
+    fn new(auth_token: &str, key: &str, url: &str) -> IdempotencyKey {
+        let mut hasher = Blake2b512::new();
+
+        hasher.update(auth_token);
+        hasher.update(":");
+        hasher.update(key);
+        hasher.update(":");
+        hasher.update(url);
+
+        let res = hasher.finalize();
+        IdempotencyKey(base64::encode(&res))
+    }
+}
+
+impl CacheKey for IdempotencyKey {
+    const PREFIX_CACHE: &'static str = "SVIX_IDEMPOTENCY_CACHE";
+}
+
+impl AsRef<str> for IdempotencyKey {
+    fn as_ref(&self) -> &str {
+        &self.0
+    }
+}
+
+async fn resolve_service<S>(
+    mut service: S,
+    req: Request<Body>,
+) -> Result<Response<BoxBody>, Infallible>
+where
+    S: Service<Request<Body>, Error = Infallible> + Clone + Send + 'static,
+    S::Response: IntoResponse,
+    S::Future: Send + 'static,
+{
+    service.call(req).await.map(IntoResponse::into_response)
+}
+
+/// The idempotency middleware itself -- inserted in place of a regular route
+#[derive(Clone)]
+pub struct IdempotencyService<S: Clone> {
+    pub redis: Option<RedisCache>,
+    pub service: S,
+}
+
+impl<S> Service<Request<Body>> for IdempotencyService<S>
+where
+    S: Service<Request<Body>, Error = Infallible> + Clone + Send + 'static,
+    S::Response: IntoResponse,
+    S::Future: Send + 'static,
+{
+    type Response = Response<BoxBody>;
+    type Error = Infallible;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        let mut service = self.service.clone();
+        let redis = self.redis.clone();
+
+        if let Some(redis) = redis {
+            Box::pin(async move {
+                let (parts, body) = req.into_parts();
+
+                // If not a POST request, simply resolve the service as usual
+                if parts.method != http::Method::POST {
+                    return resolve_service(service, Request::from_parts(parts, body)).await;
+                }
+
+                let auth = if let Some(Ok(auth)) =
+                    parts.headers.get("Authorization").map(|v| v.to_str())
+                {
+                    auth
+                } else {
+                    // No auth token -- pass off to service and do not cache
+                    return resolve_service(service, Request::from_parts(parts, body)).await;
+                };
+
+                let key = if let Some(Ok(key)) =
```

Comment:
Tiny tiny comment, but: move this if above the auth if. The reason is that this if is much more likely to be false and cause an early return, so may as well have it before auth (but after the POST check).

It doesn't really matter much.